Remove URI reserved characters from password generator

timheuer opened the pull request:
Fixes #2321 by removing URI reserved characters.

This would change it globally so that any resource builder using the util doesn't have to worry about potential server/URI conflicts with special characters reserved for server/URI host names.  Other potential is to just use no special characters in the places where this might be an issue (e.g., RabbitMQ where I discovered it). 

This PR takes the global approach.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2323)

timheuer committed:
Remove URI reserved characters from password generator

eerhardt commented:
> Other potential is to just use no special characters in the places where this might be an issue (e.g., RabbitMQ where I discovered it).

I think this approach makes a lot of sense. Limiting our special characters for all cases seems less desirable since it will create weaker passwords.

cc @sebastienros

sebastienros commented:
Interesting, I remenber I checked the RabbitMQ doc to find how these were limited and didn't find anything, but didn't think about it being passed in a url.

I think it's ok to limit the list. Though ideally, we should not have to do it but use encoding otherwise we'll always find something that prevents a specific char. That's why there is a helper method to encode `=` when used in connection strings.

If the case where a password is injected in a url it should be url-encoded, and maybe we should have some tests that enforce that we are doing it. If the password was coming from and ENV, would this issue still happen with our current code?

sebastienros commented:
For now I also think it's better to just use alphanumeric in the generated password. Less questions that could block merging. In parallel we should look at encoding all passwords.

timheuer committed:
Modifying to only gen-delims and adding test

timheuer commented:
Modified it to only restricting the gen-delims group which is less.  Added a test case and ran that test iteration 1000 times and all passed.  So the reduced character set is now minimized to only the gen-delims group in the RFC.  Removed only those `[]@/:`

Would you consider this approach? For URIs it is now valid. 

>password is injected in a url it should be url-encoded

`System.Uri` at least in .NET is doing this for us when creating a new instance with the string.

timheuer committed:
Modified test PR feedback

timheuer commented on tests/Aspire.Hosting.Tests/Utils/PasswordGeneratorTests.cs:
```diff
@@ -80,4 +88,22 @@ public void IncludesAll(int length)
+        var password = PasswordGenerator.GeneratePassword(length, length, length, length);
+        var fakeUri = new Uri($"https://guest:{password}@localhost:12345");
+
+        _output.WriteLine($"Generated password: {password}");
```
Modified. The System.Uri parser actually does the validation so I modified to ensure no exception happens in that validation.

eerhardt commented:
@timheuer - this is still in draft. Is it ready for review (and merge)? We hit this bug in our end-to-end tests we are trying to get running in CI - which shows that we really need those tests running.

timheuer commented:
undrafted it :-) @eerhardt
